README.md

```diff
@@ -161,23 +161,24 @@ any errors encountered during the request.
 
 If the server's response contains a body it MUST follow the requirements for [GraphQL response](https://graphql.github.io/graphql-spec/June2018/#sec-Response).
 
-Note: For any non-2XX response, the client should not rely on the body to be in GraphQL format.
+Note: For any non-`200` response, the client should not rely on the body to be in GraphQL format.
 The source of the response may not be the GraphQL server, but instead some intermediary such as API gateways, firewalls, etc.
 
 ## Status Codes
 
-The result of executing a GraphQL operation may contain partial data as well as encountered errors.
-Errors that happen during execution of the GraphQL operation become part of the result.
-
 If the server successfully executed parts of the operation and returns a well-formed response
-that contains a non-null key `data`, the status code SHOULD be `2xx`.
+that contains a non-null key `data`, the status code MUST be `200`.
 
 If the status code is `200`, the response MUST be a well-formed GraphQL response.
 
-The server SHOULD respond with status code `400` (Bad Request) in case of a validation error that
-prevented execution of the request.
+The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation become part of the result.
+
+In case of errors that completely prevent the successful execution of the request, the server SHOULD respond
+with the appropriate status code depending on the concrete error condition. For example:
 
-The server SHOULD respond with the appropriate status code depending on the concrete error condition,
-for example `401` (Unauthorized) or `500` (Internal Server Error).
+- `400` (Bad Request), in case of a validation error
```

Comment:
I believe that in our last group conversation, we did not reach consensus on whether validation errors should respond with a 400.

Reply:
This example section is prefaced with

> In case of errors that completely prevent the successful execution of the request

There might be validation errors that only cause a partial error, those would not fall into this category of errors.

Reply:
What validation errors would only cause a partial error? Is that `execute`-time validation of variables, perhaps?

Reply:
```graphql
mutation turnUpTheHeat(degrees: 100)
```

```json
{
  "data": {
    "turnUpTheHeat": 80
  },
  "errors": [{ "message": "That's too hot", ... }]
}
```

Reply:
The way I see it, we have three different types of validation errors:

1. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
2. Custom document validations that can be performed before execution (e.g. [graphql-depth-limit](https://github.com/stems/graphql-depth-limit/))
3. Runtime validations performed by resolvers (i.e. during the execution of the GraphQL operation).

I group 1 and 2 together, and those should result in a 400.

Type 3 should NOT result in a 400, these are errors that occur during execution, so it's a partial success. Other fields in the query could execute independently and successfully, so using a 400 status goes against the partial-success ideal of GraphQL, and determining if there _were_ any additional fields seems beyond the scope of this spec.

(Aside: I edited the word "query" out of this comment in so many places! :sweat_smile:)

Reply:
Maybe we can summarise this by changing "validation" to "GraphQL document validation"?

```suggestion
- `400` (Bad Request), in case of a GraphQL document validation error
```

Reply:
There is another type (maybe type 0?), for instance malformed JSON or typos like `{"qurey":...`. Those don't have anything to do with GraphQL documents.

The discussion we are having shows there is plenty of room for ambiguity and leaves implementors guessing. We could at least add a non-normative example of how different validation errors might be handled, akin to the example for 403.

Reply:
You're totally right. I've added type 0 to my comment above.

Reply:
Included the error types within the spec text.